Python/compile.c

```diff
@@ -354,10 +357,22 @@ if (0) {
         default:
             Py_UNREACHABLE();
     }
-    while (caches--) {
+    if (instruction->i_opcode == BINARY_OP_R) {
+        caches--;  /* last cache entry is used for the op */
+    }
+    while (caches > 0) {
         codestr->opcode = CACHE;
         codestr->oparg = 0;
         codestr++;
+        caches--;
+    }
+    if (instruction->i_opcode == BINARY_OP_R) {
+        fprintf(stderr, "instruction->i_cache_data = %d\n", instruction->i_cache_data);
+        assert(instruction->i_cache_data >= 0);
+        assert(instruction->i_cache_data < 256);
+        codestr->opcode = instruction->i_cache_data & 0xFF;
+        codestr->oparg = 0;
+        codestr++;
```

Comment:
@markshannon, @gvanrossum Is this what you had in mind? I'm not sure I understand the implications of using the cache in this way. I could make OPSIZE take an opcode and return 3 instead of 2 for BINARY_OP_R. That wouldn't be hard to do. What do you think?

(This version works for me on an isolated test, but there is still some problem getting a build when I enable it globally.)

Reply:
Not all instructions need take two code units. `RETURN_VALUE`, `YIELD_VALUE`, RESUME and some others will fit into one code unit.

We already have a table for cache sizes, so it shouldn't be too hard to add an `Opsize` table as well.
That way you won't need to special case `BINARY_OP_R`
The code can become:
```C
int opsize = _PyOpcode_Opsize[i->opcode];
switch (opsize) {
    case 3:
        codestr[2].cache = instruction->i_cache_data;
        /* fall through */
    case 2:
        codestr[1].oparg2 = oparg2 & 0xFF;
        codestr[1].oparg3 = oparg3 & 0XFF;
        /* fall through */
    case 1:
        codestr[0].opcode = opcode;
        codestr[0].oparg = oparg1 & 0xFF;
        break;
    default:
        ERROR!
}
codestr += opsize;
```

Reply:
Yes, I'm trying that now. The's a small complication where we have a prev_instr that could be pointing to a cache entry, and we increment it -  used to be by 1, not it's by 2, and if we want it to depend on the opcode then we need to have the opcode somewhere). 

There's one in _PyEval_EvalFrameDefault, which is currently:

`next_instr = frame->prev_instr + (_PyInterpreterFrame_LASTI(frame) == -1 ? 1 : OPSIZE); `

and another in genobject.c, _PyGen_yf:

`_Py_CODEUNIT next = frame->prev_instr[OPSIZE];`

Reply:
Also in DISPATCH_INLINED we have  `frame->prev_instr = next_instr - OPSIZE; `

Reply:
I could revert the change I made in eval.c, so that INSTRUCTION_START etc always do `next_instr++` instead of `next_instr += OPSIZE(op);`, and each instruction does a `JUMPBY(CACHES + OPSIZE -1)`.